crates/deltalake-core/src/operations/merge.rs

```diff
@@ -652,6 +657,237 @@ impl ExtensionPlanner for MergeMetricExtensionPlanner {
+                partition_columns,
+                source_name,
+                target_name,
+                placeholders,
+            );
+            let right = generalize_filter(
+                *binary.right,
+                partition_columns,
+                source_name,
+                target_name,
+                placeholders,
+            );
+
+            match (left, right) {
+                (None, None) => None,
+                (None, Some(r)) => Some(r),
+                (Some(l), None) => Some(l),
+                (Some(l), Some(r)) => Expr::BinaryExpr(BinaryExpr {
+                    left: l.into(),
+                    op: binary.op,
+                    right: r.into(),
+                })
+                .into(),
+            }
+        }
+        other => Some(other),
+    }
+}
+
+fn replace_placeholders(expr: Expr, placeholders: &HashMap<String, ScalarValue>) -> Expr {
+    expr.transform(&|expr| match expr {
+        Expr::Placeholder(Placeholder { id, .. }) => {
+            let value = placeholders[&id].clone();
+            // Replace the placeholder with the value
+            Ok(Transformed::Yes(Expr::Literal(value)))
+        }
+        _ => Ok(Transformed::No(expr)),
+    })
+    .unwrap()
+}
+
+async fn try_construct_early_filter(
+    join_predicate: Expr,
+    table_snapshot: &DeltaTableState,
+    session_state: &SessionState,
+    source: &LogicalPlan,
+    source_name: &TableReference<'_>,
+    target_name: &TableReference<'_>,
+) -> Option<Expr> {
+    let table_metadata = table_snapshot.metadata()?;
+
+    let partition_columns = &table_metadata.partition_columns;
+
+    if partition_columns.is_empty() {
+        return None;
+    }
+
+    let mut placeholders = HashMap::default();
+
+    let filter = generalize_filter(
+        join_predicate,
+        partition_columns,
+        source_name,
+        target_name,
+        &mut placeholders,
+    )?;
+
+    if placeholders.is_empty() {
+        // if we haven't recognised any partition-based predicates in the join predicate, return our reduced filter
+        Some(filter)
+    } else {
+        // if we have some recognised partitions, then discover the distinct set of partitions in the source data and
+        // make a new filter, which expands out the placeholders for each distinct partition (and then OR these together)
+        let distinct_partitions = LogicalPlan::Distinct(Distinct {
+            input: LogicalPlan::Projection(
+                Projection::try_new(
+                    placeholders
+                        .into_iter()
+                        .map(|(alias, expr)| expr.alias(alias))
+                        .collect_vec(),
+                    source.clone().into(),
+                )
+                .unwrap(),
+            )
+            .into(),
+        });
+
+        let execution_plan = session_state
+            .create_physical_plan(&distinct_partitions)
+            .await
+            .unwrap();
```

Comment:
There are quite a few unwraps here and there doesn't seem to be an guarantee they won't panic. Use ? if possible or provide a generic delta table error